### ISSUE TITLE
SDPA kernel

### ISSUE THREAD

**Issue by cglagovichTT** [tenstorrent/tt-metal#7461]
For Llama2 prefill of prompt length 2k, SPDA is currently implemented with 8 passes of partial interleaving, matmul -> softmax -> matmul. This module takes 2.5ms, and constitutes 12% of prefill runtime. We think that by implementing a FlashAttention-style SDPA kernel, we can remove the overhead of partial sharding and reduce the execution time of SDPA.

- [x] Python implementation of FlashAttention2
- [x] single-core kernel skeleton and host skeleton
- [x] single-core kernel correctness
- [x] multi-core kernel correctness
- [x] optimized causal attention (don't compute scores for cols > rows, don't read full mask, only apply to block where cols == rows)
- [ ] optimized K, V reading/sharing (multi-cast?)

**Comment by cglagovichTT** [comment]
Currently, a version of the kernel where it only does blockwise `Q @ K` and `SCORES @ V` is functionally correct when compared to pytorch reference.

Perf measurements show that on single core, this runs in 63 ms. When removing the matmul calls, this runs in 4 ms. 

Reem recommends the following pattern to optimize matmuls:
```
tile_regs_acquire();
math call
tile_regs_commit();

tile_regs_wait();
pack call 
tile_regs_release(); 
```

**Comment by cglagovichTT** [comment]
With SDPA functional on a single core, this is what we're seeing in terms of perf.

<img width="836" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/114512306/1b6bbb4f-ee79-46fc-82d7-9e6243fc2bd0">

**Comment by cglagovichTT** [comment]
After implementing naive parallelization over q_heads=8 and q_chunks=8 on 64 cores, I get 49x speedup over single core implementation. 


```
    b = 1
    nh = 8
    nkv = 1
    s = 2048
    d = 128
```
FW duration = 1.77 ms, faster than unfused 2.5 ms.

**Comment by cglagovichTT** [comment]
Let's assert that `batch * q_heads` divides `grid_size`. That way we know that each core is never responsible for more than one `batch` or one `q_head`.

For core `i` of `num_cores`, with `batch`, `q_heads`, `q_parallel_factor`, a core will do the following.

```
num_local_q_chunks = num_q_chunks // q_parallel_factor
my_batch = i // (q_heads * q_parallel_factor)
my_q_head = (i // q_parallel_factor) % q_heads
my_q_chunk_start = i % q_heads
```

From this, a core can loop from `my_q_chunk_start` to `my_q_chunk_start + num_local_q_chunks` and calculate all necessary indices.

**Comment by cglagovichTT** [comment]
Reading attn_mask accounts for 13.56% of the runtime. To solve:

1. Don't compute Q * K when k_chunk_idx > q_chunk_idx
2. Assume causal attention, so only read attn_mask if k_chunk_idx == q_chunk_idx

**Comment by cglagovichTT** [comment]
That helped by 1.2x

**Comment by cglagovichTT** [comment]
Update:

I added support such that each core is capable of computing more than one q_chunk. For now, if I have 8 q_chunks and I make the q_parallel_factor=4, then each core computes the result for 2 contiguous q_chunks.

In other news, @rtawfik01 found non-determinism to be caused by the following methodology:
```python
for i in range(num_tiles):
  cb_wait_front(in0, 1);
  do_math(in0, in1);
  cb_pop_front(in0, 1);
  cb_reserve_back(in0, 1);
  pack_tile(0, in0);
  cb_push_back(in0, 1);
```
This pattern is a problem since it compute an in-place operation on in0 which should have `num_tiles` produced at the beginning. The first `cb_wait_front(in0, 1);` always passes since there is always `num_tiles - 1` produced in the CB. The fix that she found is to `cb_wait_front(in0, num_tiles);` in the inner loop to ensure that the `cb_push_back` has completed before progressing.

**Comment by cglagovichTT** [comment]
This is the tracy for the longest core 
<img width="2572" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/114512306/d8ae7a60-50b4-4455-b6b2-d9339bf84f40">


This core loops over more k_chunks than any other core, so it takes 1.46 ms. To compare, the core which does the least work is done in .2 ms. This happens because in causal attention, q_chunk 0 only attends to k_chunk 0. q_chunk 7 attends to k_chunk 0 through 7. 

Current parallelization looks like this 
<img width="1463" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/114512306/86c4edc7-b7c6-4056-9b8a-4fc636797814">


If I can split the Q chunks further such that each core is responsible for 2 Q chunks, then I can make each core work on a short one and a long one like this.

![Uploading image.png…]()

**Comment by cglagovichTT** [comment]
there are 8 cores which each only use 1 K chunk. These cores range from 199 us to 312 us. Looks like the delay is for data incoming.

There are 8 cores which use all 8 K chunks. These cores range from 1.4 ms to 1.46 ms.

Right now, each inner-loop is 166 us. If I split q chunks by 2, each inner loop should be about 83 us (this is very hopeful, since we will now be reading K and V twice as often).

This parallelization scheme should make each core run in about 0.73 ms in the worst case. Note that it is only valid for causal attention, but that is the main use case for this SDPA kernel.

**Comment by cglagovichTT** [comment]
I added a feature that allows q_chunk and k_chunk to be different. This slightly complicated the loop bounds since if q_chunk = 4 tiles and k_chunk = 8 tiles, then q_chunk 0 and q_chunk 1 both use up to k_chunk 0. and both q_chunk 0 and q_chunk 1 need to read the attn_mask when operating on k_chunk 0

**Comment by cglagovichTT** [comment]
I found that when we fixed the `cb_wait_front` to wait for full input, we hurt our perf. Went from 1.4ms to 1.65ms. 

I also checked out the fast exp commit [74b5bc7bd47a770cdde4cc1547eebff8c20c5745](https://github.com/tenstorrent/tt-metal/commit/74b5bc7bd47a770cdde4cc1547eebff8c20c5745) and got poor PCC, Reem is looking into it.

I tried out the better parallelization scheme and got time down from 1.65 ms to 1 ms. The implementation was very hacky so I put it in different files.

**Comment by cglagovichTT** [comment]
Latest update: SDPA can choose parallelization factors over batch, heads, and q sequence. 

We still get ND PCC when we try a variety of optimizations:
- granularity 8 in `mul_block_bcast_scalar_inplace`
- unfuse `sub_exp_block`
- remove unnecessary `cb_wait_front` in `mul_block_bcast_cols_inplace`

**Comment by cglagovichTT** [comment]
Llama prefill 2k now runs in `0.414828 ms`. Removing reader/writer leads to `.35 ms`. Appears that there is still a compute-bound portion of the kernel. Optimizing out a few eltwise adds by accumulating in dest might be the best way to squeeze out the last bit of perf.


For batch 1, SDPA attains `41.4` TFLOPs/s calculated as `2 * seq**2 * n_heads * head_dim * batch * 2 (HiFi2)`. 
At batch 8, SDPA attains `51.0` TFLOPs/s. Compare to FlashAttention-2/s 187 TFLOPs/s on A100.

**Comment by cglagovichTT** [comment]
Merged in #7886
